### DISCUSSION TITLE
complex `if-*` operations

### DISCUSSION THREAD

**Discussion by jhheider** [pkgxdev/brewkit#288] [Ideas]
since we can't do things like `if: darwin && ^3` currently in build scripts, my brain has considered the following as options:
```yaml
  - run: foo
    if-all: [darwin, ^4]
  - run: bar
    if-any: [linux, >3.2.1]
```
it's needed somewhat infrequently, but would solve things like:
```yaml
    # Building compiler-rt on darwin+aarch64 fails for versions less than
    # 14 with the below configuration. FIXME if possible, of course.
    - run: |
        if test "{{hw.platform}}" = "linux"; then
          ARGS="$ARGS -DLLVM_ENABLE_RUNTIMES='compiler-rt'"
        elif semverator satisfies '>=14' {{version}}; then
          ARGS="$ARGS -DLLVM_ENABLE_RUNTIMES='compiler-rt'"
        elif test "{{hw.arch}}" = "x86-64" && semverator satisfies '>=14' {{version}}; then
          ARGS="$ARGS -DLLVM_ENABLE_RUNTIMES='compiler-rt'"
        fi
      # if: linux || >=14 || (x86-64 && >=13)  ## Brewkit can't do this. Yet.
```
somewhat.

care would have to be taken to ensure that `if:`, `if-any:`, and `if-all:` were mutually exclusive. potentially `if-not:` and `if-none:` could also be useful.

**Comment by mxcl** [comment]
I feel like we should avoid inventing syntax for YAML. YAML is not a programming language, so probs rather than end up like GitHub Actions we should just have you write a shell script `if` for more complexity.

Thoughts?

**Reply by mxcl** [reply]
OTOH it's pretty common to need a branch for a platform version range so that seems fine and then it's always `AND` in an array and if you need any you have to do a shell `if`?

**Reply by jhheider** [reply]
also feasible. will result in more text, but it's more flexible and easier to document. i always find boolean logic a natural way to express stuff.

if it just comes down to shell scripting, then what we have is what we want already.

**Reply by mxcl** [reply]
I guess my main beef with github actions is they support `==` and that but it's not intuitive what will work or not because it's just a basic parser and not a real language.

If we *only* support && and || then it may be ok. But then there’s the need for parentheses…

**Reply by jhheider** [reply]
right, it's why my idea had evolved into moving the logic out to the key name `if-any`, etc. reduces parsing of the key value.

i _agree_ that we don't want to have to maintain gha-style complexity.
